File: cli/core/leitner.go
// Projeção Leitner (spec/SPEC.md §4.1) — RF08/RF11.
// 5 caixas; acerto promove, erro regride UMA caixa (variante atenuada).
package core

import (
	"cmp"
	"slices"
)

const maxBox = 5

// intervalDays[caixa] = dias até a próxima revisão (índice 0 não é usado).
var intervalDays = [maxBox + 1]int{0, 1, 3, 7, 14, 30}

type LeitnerCard struct {
	DeckID           string   `json:"deck_id"`
	Box              int      `json:"box"`
	Due              string   `json:"due"`
	Front            string   `json:"front"`
	Back             string   `json:"back"`
	SourceURL        *string  `json:"source_url"`
	SourceTitle      *string  `json:"source_title"`
	Tags             []string `json:"tags"`
	LastExplanation  *string  `json:"last_explanation"`
	ExplanationCount int      `json:"explanation_count"`
	LastFrame        string   `json:"last_frame"` // último andaime; ausente ≡ "feynman"
	Archived         bool     `json:"archived"`
}

type LeitnerState struct {
	Cards map[string]LeitnerCard `json:"cards"`
	Queue []string               `json:"queue"`
}

func init() {
	Register("leitner", func(events []Event, p Params) (any, error) {
		return Leitner(events, p)
	})
}

func Leitner(events []Event, p Params) (LeitnerState, error) {
	type cardState struct {
		LeitnerCard
		id        string
		createdTS string
	}
	cards := map[string]*cardState{}
	var order []*cardState // ordem de criação no log normalizado
	archivedDecks := map[string]bool{}

	for _, event := range Normalize(events) {
		switch event.Type {
		case "card.created":
			pl, err := decodePayload[struct {
				CardID      string   `json:"card_id"`
				DeckID      string   `json:"deck_id"`
				Front       string   `json:"front"`
				Back        string   `json:"back"`
				SourceURL   *string  `json:"source_url"`
				SourceTitle *string  `json:"source_title"`
				Tags        []string `json:"tags"`
			}](event)
			if err != nil {
				return LeitnerState{}, err
			}
			due, err := LocalDate(event.TS, p.TZ)
			if err != nil {
				return LeitnerState{}, err
			}
			card := &cardState{
				LeitnerCard: LeitnerCard{
					DeckID:      pl.DeckID,
					Box:         1,
					Due:         due,
					Front:       pl.Front,
					Back:        pl.Back,
					SourceURL:   pl.SourceURL,
					SourceTitle: pl.SourceTitle,
					Tags:        orEmpty(pl.Tags),
					LastFrame:   "feynman",
				},
				id:        pl.CardID,
				createdTS: event.TS,
			}
			cards[pl.CardID] = card
			order = append(order, card)

		case "card.edited":
			pl, err := decodePayload[struct {
				CardID string    `json:"card_id"`
				Front  *string   `json:"front"`
				Back   *string   `json:"back"`
				Tags   *[]string `json:"tags"`
			}](event)
			if err != nil {
				return LeitnerState{}, err
			}
			card, ok := cards[pl.CardID]
			if !ok {
				continue
			}
			// last-writer-wins por campo; caixa e vencimento não mudam
			if pl.Front != nil {
				card.Front = *pl.Front
			}
			if pl.Back != nil {
				card.Back = *pl.Back
			}
			if pl.Tags != nil {
				card.Tags = orEmpty(*pl.Tags)
			}

		case "card.archived":
			pl, err := decodePayload[struct {
				CardID string `json:"card_id"`
			}](event)
			if err != nil {
				return LeitnerState{}, err
			}
			if card, ok := cards[pl.CardID]; ok {
				card.Archived = true
			}

		case "deck.archived":
			pl, err := decodePayload[struct {
				DeckID string `json:"deck_id"`
			}](event)
			if err != nil {
				return LeitnerState{}, err
			}
			archivedDecks[pl.DeckID] = true

		case "card.explained":
			pl, err := decodePayload[struct {
				CardID string `json:"card_id"`
				Text   string `json:"text"`
				Frame  string `json:"frame"` // v2; ausente ≡ "feynman"
			}](event)
			if err != nil {
				return LeitnerState{}, err
			}
			card, ok := cards[pl.CardID]
			if !ok {
				continue // explicação de cartão inexistente é ignorada
			}
			text := pl.Text
			card.LastExplanation = &text
			card.ExplanationCount++
			if pl.Frame != "" {
				card.LastFrame = pl.Frame
			} else {
				card.LastFrame = "feynman"
			}

		case "card.reviewed":
			pl, err := decodePayload[struct {
				CardID string `json:"card_id"`
				Result string `json:"result"`
			}](event)
			if err != nil {
				return LeitnerState{}, err
			}
			card, ok := cards[pl.CardID]
			if !ok {
				continue
			}
			if pl.Result == "correct" {
				card.Box = min(card.Box+1, maxBox)
			} else {
				card.Box = max(card.Box-1, 1)
			}
			reviewDay, err := LocalDate(event.TS, p.TZ)
			if err != nil {
				return LeitnerState{}, err
			}
			card.Due, err = AddDays(reviewDay, intervalDays[card.Box])
			if err != nil {
				return LeitnerState{}, err
			}
		}
	}

	today, err := LocalDate(p.Now, p.TZ)
	if err != nil {
		return LeitnerState{}, err
	}

	var due []*cardState
	for _, card := range order {
		if card.Due <= today && !card.Archived && !archivedDecks[card.DeckID] {
			due = append(due, card)
		}
	}
	slices.SortStableFunc(due, func(a, b *cardState) int {
		if a.Box != b.Box {
			return a.Box - b.Box
		}
		if a.Due != b.Due {
			return cmp.Compare(a.Due, b.Due)
		}
		return cmp.Compare(a.createdTS, b.createdTS)
	})

	state := LeitnerState{Cards: make(map[string]LeitnerCard, len(cards)), Queue: []string{}}
	for id, card := range cards {
		state.Cards[id] = card.LeitnerCard
	}
	for _, card := range due {
		state.Queue = append(state.Queue, card.id)
	}
	return state, nil
}

File: extension/core/leitner.js
// Projeção Leitner (spec/SPEC.md §4.1) — RF08/RF11.
// 5 caixas; acerto promove, erro regride UMA caixa (variante atenuada).

import { normalize } from "./envelope.js";
import { localDate, addDays } from "./time.js";

const MAX_BOX = 5;
const INTERVAL_DAYS = { 1: 1, 2: 3, 3: 7, 4: 14, 5: 30 };

export function project(events, { now, tz }) {
  const cards = new Map();
  const archivedDecks = new Set();

  for (const event of normalize(events)) {
    if (event.type === "card.created") {
      const { card_id, deck_id, front, back, source_url, source_title, tags } =
        event.payload;
      cards.set(card_id, {
        deck_id,
        box: 1,
        due: localDate(event.ts, tz),
        createdTs: event.ts,
        front,
        back,
        source_url: source_url ?? null,
        source_title: source_title ?? null,
        tags: tags ?? [],
        last_explanation: null,
        explanation_count: 0,
        last_frame: "feynman", // último andaime; ausente ≡ "feynman"
        archived: false,
      });
    } else if (event.type === "card.edited") {
      const card = cards.get(event.payload.card_id);
      if (!card) continue;
      // last-writer-wins por campo; caixa e vencimento não mudam
      const { front, back, tags } = event.payload;
      if (front !== undefined) card.front = front;
      if (back !== undefined) card.back = back;
      if (tags !== undefined) card.tags = tags ?? [];
    } else if (event.type === "card.archived") {
      const card = cards.get(event.payload.card_id);
      if (card) card.archived = true;
    } else if (event.type === "deck.archived") {
      archivedDecks.add(event.payload.deck_id);
    } else if (event.type === "card.explained") {
      const card = cards.get(event.payload.card_id);
      if (!card) continue; // explicação de cartão inexistente é ignorada
      card.last_explanation = event.payload.text;
      card.explanation_count += 1;
      card.last_frame = event.payload.frame ?? "feynman";
    } else if (event.type === "card.reviewed") {
      const card = cards.get(event.payload.card_id);
      if (!card) continue;
      card.box =
        event.payload.result === "correct"
          ? Math.min(card.box + 1, MAX_BOX)
          : Math.max(card.box - 1, 1);
      card.due = addDays(localDate(event.ts, tz), INTERVAL_DAYS[card.box]);
    }
  }

  const today = localDate(now, tz);
  const queue = [...cards.entries()]
    .filter(
      ([, card]) =>
        card.due <= today && !card.archived && !archivedDecks.has(card.deck_id)
    )
    .sort(([, a], [, b]) => {
      if (a.box !== b.box) return a.box - b.box;
      if (a.due !== b.due) return a.due < b.due ? -1 : 1;
      return a.createdTs < b.createdTs ? -1 : 1;
    })
    .map(([cardId]) => cardId);

  const cardsOut = {};
  for (const [cardId, card] of cards) {
    const {
      deck_id, box, due, front, back, source_url, source_title, tags,
      last_explanation, explanation_count, last_frame, archived,
    } = card;
    cardsOut[cardId] = {
      deck_id, box, due, front, back, source_url, source_title, tags,
      last_explanation, explanation_count, last_frame, archived,
    };
  }
  return { cards: cardsOut, queue };
}
